cpu/sam0_common/usb/usb.c

```diff
@@ -0,0 +1,679 @@
+/*
+ * Copyright (C) 2019 Koen Zandberg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup sam0_common
+ * @{
+ * @file
+ * @brief   USB interface functions for the sam0 class devices
+ *
+ * @author  Koen Zandberg <koen@bergzand.net>
+ * @}
+ */
+#include <stdint.h>
+#include <stdlib.h>
+#include <errno.h>
+#include "cpu.h"
+#include "periph/gpio.h"
+#include "usb/usbdev.h"
+#include "sam_usb.h"
+
+#include "bitarithm.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+static sam0_common_usb_t *_usbdev;
+
+/* Descriptor bank */
+static UsbDeviceDescBank banks[2 * SAM_USB_NUM_EP];
+static usbdev_ep_t endpoints[2 * SAM_USB_NUM_EP];
+
+const usbdev_ep_driver_t driver_ep;
+static void usbdev_ep_init(usbdev_ep_t *ep);
+static int usbdev_ep_ready(usbdev_ep_t *ep, size_t len);
+
+
+static usbdev_ep_t *usbdev_new_ep(usbdev_t *dev, usb_ep_type_t type,
+                                  usb_ep_dir_t dir, size_t buf_len);
+
+static void _ep_address(usbdev_ep_t *ep);
+static void _ep_size(usbdev_ep_t *ep);
+static int _ep_unready(usbdev_ep_t *ep);
+
+static inline unsigned _get_ep_num(unsigned num, usb_ep_dir_t dir)
+{
+    return 2 * num + (dir == USB_EP_DIR_OUT ? 0 : 1);
+}
+
+static inline unsigned _get_ep_num2(usbdev_ep_t *ep)
```

Comment:
Why is there two  get_ep_num functions ?

Reply:
Probably just bad naming, but will check.

Reply:
Refactored these a bit to make more sense